File: sdks/dart/lib/src/cheki_client.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;

import 'models.dart';
import 'verify_options.dart';

/// A client for the cheki receipt verification API.
///
/// cheki is a free Ethiopian receipt verification service. This client
/// provides typed access to the four REST endpoints:
///
/// * [verify] — `POST /api/verify`
/// * [verifyBatch] — `POST /api/verify/batch`
/// * [getBanks] — `GET /api/banks`
/// * [getHealth] — `GET /api/health`
///
/// ## Usage
///
/// ```dart
/// final client = ChekiClient();
/// final result = await client.verify(
///   bank: 'cbe',
///   reference: 'RT123456789',
///   accountNumber: '1000123456789',
/// );
/// print(result.verified);
/// client.close();
/// ```
///
/// You can override the base URL (e.g. for testing or self-hosting):
///
/// ```dart
/// final client = ChekiClient(baseUrl: 'http://localhost:3000');
/// ```
class ChekiClient {
  /// The base URL of the cheki API (no trailing slash).
  final String baseUrl;

  /// The underlying HTTP client used for requests.
  ///
  /// Exposed so callers can inject a mock client in tests.
  final http.Client httpClient;

  /// Default headers sent with every request.
  final Map<String, String> _defaultHeaders = {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    'User-Agent': 'cheki-dart-sdk/0.1.0',
  };

  /// Creates a new [ChekiClient].
  ///
  /// [baseUrl] defaults to the production cheki service.
  /// Pass a custom [client] to inject a mock or configured HTTP client.
  ChekiClient({
    String? baseUrl,
    http.Client? client,
  })  : baseUrl = (baseUrl ?? 'https://cheki-pi.vercel.app').replaceAll(
            RegExp(r'/+$'),
            ''),
        httpClient = client ?? http.Client();

  /// Verifies a single bank transfer receipt.
  ///
  /// [bank] is the bank code (e.g. `cbe`, `dashen`).
  /// [reference] is the transaction reference number.
  /// Provide [accountNumber], [phoneNumber], or [qrData] as required
  /// by the specific bank.
  ///
  /// Alternatively, pass a [VerifyOptions] object via [options] for
  /// a more structured approach.
  ///
  /// Returns a [VerifyResult] containing the verification details.
  ///
  /// Throws [ChekiException] on API-level errors or [FormatException]
  /// if the response cannot be decoded.
  Future<VerifyResult> verify({
    required String bank,
    required String reference,
    String? accountNumber,
    String? phoneNumber,
    String? qrData,
    VerifyOptions? options,
  }) async {
    final Map<String, dynamic> body;

    if (options != null) {
      body = options.toJson();
    } else {
      body = {
        'bank': bank,
        'reference': reference,
        if (accountNumber != null) 'accountNumber': accountNumber,
        if (phoneNumber != null) 'phoneNumber': phoneNumber,
        if (qrData != null) 'qrData': qrData,
      };
    }

    final response = await _post('/api/verify', body);
    final json = _decodeJson(response);
    return VerifyResult.fromJson(json);
  }

  /// Verifies multiple receipts in a single batch request.
  ///
  /// [receipts] is a list of [BatchReceipt] objects, each containing
  /// the bank code, reference, and optional account/phone numbers.
  ///
  /// Returns a [BatchResult] with aggregate counts and per-receipt results.
  ///
  /// Throws [ChekiException] on API-level errors.
  Future<BatchResult> verifyBatch(List<BatchReceipt> receipts) async {
    final body = {
      'receipts': receipts.map((r) => r.toJson()).toList(),
    };

    final response = await _post('/api/verify/batch', body);
    final json = _decodeJson(response);
    return BatchResult.fromJson(json);
  }

  /// Retrieves the list of banks supported by cheki.
  ///
  /// Returns a list of [BankInfo] objects.
  ///
  /// Throws [ChekiException] on API-level errors.
  Future<List<BankInfo>> getBanks() async {
    final response = await _get('/api/banks');
    final json = _decodeJson(response);

    final banksRaw = json['banks'] as List?;
    if (banksRaw == null) {
      throw ChekiException(
        'Unexpected response from /api/banks: missing "banks" field',
        statusCode: response.statusCode,
      );
    }

    return banksRaw
        .map((e) => BankInfo.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Checks the health of the cheki service.
  ///
  /// Returns a [HealthStatus] with the service status, version, and
  /// individual check results.
  ///
  /// Throws [ChekiException] on API-level errors.
  Future<HealthStatus> getHealth() async {
    final response = await _get('/api/health');
    final json = _decodeJson(response);
    return HealthStatus.fromJson(json);
  }

  /// Closes the underlying HTTP client and releases resources.
  ///
  /// Call this when you are done with the client, or use the client
  /// within a try/finally block:
  ///
  /// ```dart
  /// final client = ChekiClient();
  /// try {
  ///   // use client...
  /// } finally {
  ///   client.close();
  /// }
  /// ```
  void close() {
    httpClient.close();
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  Future<http.Response> _post(
    String path,
    Map<String, dynamic> body,
  ) async {
    final uri = Uri.parse('$baseUrl$path');
    try {
      final response = await httpClient.post(
        uri,
        headers: _defaultHeaders,
        body: jsonEncode(body),
      );
      _checkStatus(response);
      return response;
    } on http.ClientException catch (e) {
      throw ChekiException(
        'HTTP request failed: ${e.message}',
        path: path,
      );
    } on SocketException catch (e) {
      throw ChekiException(
        'Network error: ${e.message}',
        path: path,
      );
    }
  }

  Future<http.Response> _get(String path) async {
    final uri = Uri.parse('$baseUrl$path');
    try {
      final response = await httpClient.get(uri, headers: _defaultHeaders);
      _checkStatus(response);
      return response;
    } on http.ClientException catch (e) {
      throw ChekiException(
        'HTTP request failed: ${e.message}',
        path: path,
      );
    } on SocketException catch (e) {
      throw ChekiException(
        'Network error: ${e.message}',
        path: path,
      );
    }
  }

  void _checkStatus(http.Response response) {
    if (response.statusCode < 200 || response.statusCode >= 300) {
      String? apiError;
      try {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        apiError = json['error'] as String?;
      } catch (_) {
        // Response body is not JSON; use raw body.
        apiError = response.body.isNotEmpty ? response.body : null;
      }
      throw ChekiException(
        apiError ?? 'HTTP ${response.statusCode}',
        statusCode: response.statusCode,
      );
    }
  }

  Map<String, dynamic> _decodeJson(http.Response response) {
    try {
      final decoded = jsonDecode(response.body);
      if (decoded is! Map<String, dynamic>) {
        throw ChekiException(
          'Unexpected response format: expected a JSON object',
          statusCode: response.statusCode,
        );
      }
      return decoded;
    } on FormatException catch (e) {
      throw ChekiException(
        'Failed to decode JSON response: ${e.message}',
        statusCode: response.statusCode,
      );
    }
  }
}

/// An exception thrown by the cheki client when an API request fails.
class ChekiException implements Exception {
  /// A human-readable error message.
  final String message;

  /// The HTTP status code associated with the error, if applicable.
  final int? statusCode;

  /// The API path that was called, if applicable.
  final String? path;

  /// Constructs a [ChekiException].
  const ChekiException(
    this.message, {
    this.statusCode,
    this.path,
  });

  @override
  String toString() {
    final parts = <String>[message];
    if (statusCode != null) parts.add('status: $statusCode');
    if (path != null) parts.add('path: $path');
    return 'ChekiException(${parts.join(', ')})';
  }
}

File: sdks/php/src/ChekiClient.php
<?php

declare(strict_types=1);

namespace Cheki;

/**
 * Main client for the cheki receipt verification API.
 *
 * Usage:
 *
 *   $client = new ChekiClient();
 *   $result = $client->verify('cbe', 'RB1234567890', '1000123456789');
 *   if ($result->isVerified()) {
 *       echo $result->senderName . ' sent ' . $result->amount . ' ' . $result->currency;
 *   }
 */
class ChekiClient
{
    /** Maximum number of retry attempts for transient failures. */
    private const MAX_RETRIES = 3;

    /** Base delay in milliseconds for exponential backoff. */
    private const BASE_BACKOFF_MS = 200;

    private VerifyOptions $options;

    public function __construct(?VerifyOptions $options = null)
    {
        $this->options = $options ?? new VerifyOptions();
    }

    /**
     * Create a client with a custom base URL.
     */
    public static function withBaseUrl(string $baseUrl): self
    {
        return new self((new VerifyOptions())->withBaseUrl($baseUrl));
    }

    /**
     * Get the current options (immutable clone pattern).
     */
    public function getOptions(): VerifyOptions
    {
        return $this->options;
    }

    // ─── Public API ───────────────────────────────────────────────

    /**
     * Verify a single receipt.
     *
     * @param string      $bank          Bank code (e.g. 'cbe', 'dashen', 'awash').
     * @param string      $reference     Receipt reference number.
     * @param string|null $accountNumber Optional account number.
     * @param string|null $phoneNumber   Optional phone number.
     * @param string|null $qrData        Optional raw QR data string.
     */
    public function verify(
        string $bank,
        string $reference,
        ?string $accountNumber = null,
        ?string $phoneNumber = null,
        ?string $qrData = null
    ): VerifyResult {
        $payload = [
            'bank'      => $bank,
            'reference' => $reference,
        ];

        if ($accountNumber !== null && $accountNumber !== '') {
            $payload['accountNumber'] = $accountNumber;
        }
        if ($phoneNumber !== null && $phoneNumber !== '') {
            $payload['phoneNumber'] = $phoneNumber;
        }
        if ($qrData !== null && $qrData !== '') {
            $payload['qrData'] = $qrData;
        }

        $response = $this->post('/api/verify', $payload);

        if ($response->error !== null) {
            return VerifyResult::error($response->error, $response->httpStatus);
        }

        return VerifyResult::fromResponse($response->data, $response->httpStatus);
    }

    /**
     * Verify a batch of receipts.
     *
     * @param array<int, array{bank: string, reference: string, accountNumber?: string, phoneNumber?: string}> $receipts
     */
    public function verifyBatch(array $receipts): BatchResult
    {
        // Normalize: strip null/empty optional fields
        $clean = [];
        foreach ($receipts as $receipt) {
            $item = [
                'bank'      => $receipt['bank'],
                'reference' => $receipt['reference'],
            ];
            if (isset($receipt['accountNumber']) && $receipt['accountNumber'] !== '') {
                $item['accountNumber'] = $receipt['accountNumber'];
            }
            if (isset($receipt['phoneNumber']) && $receipt['phoneNumber'] !== '') {
                $item['phoneNumber'] = $receipt['phoneNumber'];
            }
            $clean[] = $item;
        }

        $response = $this->post('/api/verify/batch', ['receipts' => $clean]);

        if ($response->error !== null) {
            return BatchResult::error($response->error, $response->httpStatus);
        }

        return BatchResult::fromResponse($response->data, $response->httpStatus);
    }

    /**
     * Get the list of supported banks.
     *
     * @return Bank[]
     */
    public function getBanks(): array
    {
        $response = $this->get('/api/banks');

        if ($response->error !== null) {
            return [];
        }

        $banks = [];
        if (isset($response->data['banks']) && is_array($response->data['banks'])) {
            foreach ($response->data['banks'] as $bankData) {
                if (is_array($bankData)) {
                    $banks[] = Bank::fromArray($bankData);
                }
            }
        }

        return $banks;
    }

    /**
     * Check the health of the cheki API.
     */
    public function getHealth(): HealthStatus
    {
        $response = $this->get('/api/health');

        if ($response->error !== null) {
            return new HealthStatus(
                success: false,
                status: 'unreachable',
                httpStatus: $response->httpStatus,
                raw: ['error' => $response->error]
            );
        }

        return HealthStatus::fromResponse($response->data, $response->httpStatus);
    }

    // ─── Internal HTTP layer ──────────────────────────────────────

    /**
     * Perform a GET request and return decoded JSON.
     *
     * @return object{data: array<string, mixed>, error: ?string, httpStatus: ?int}
     */
    private function get(string $path): object
    {
        return $this->request('GET', $path, null);
    }

    /**
     * Perform a POST request with a JSON body and return decoded JSON.
     *
     * @param array<string, mixed> $payload
     * @return object{data: array<string, mixed>, error: ?string, httpStatus: ?int}
     */
    private function post(string $path, array $payload): object
    {
        return $this->request('POST', $path, $payload);
    }

    /**
     * Core cURL request handler with retry support.
     *
     * Retries up to MAX_RETRIES times on HTTP 429 (Too Many Requests) and
     * 5xx (Server Error) responses using exponential backoff. Network-level
     * cURL errors and non-retryable HTTP status codes are returned immediately.
     *
     * @param string               $method  HTTP method (GET or POST).
     * @param string               $path    API path (e.g. '/api/verify').
     * @param array<string, mixed>|null $payload JSON body for POST.
     * @return object{data: array<string, mixed>, error: ?string, httpStatus: ?int}
     */
    private function request(string $method, string $path, ?array $payload): object
    {
        $url = $this->options->baseUrl . $path;

        $ch = curl_init($url);

        if ($ch === false) {
            return (object) [
                'data'       => [],
                'error'      => 'Failed to initialize cURL',
                'httpStatus' => null,
            ];
        }

        $headers = array_merge([
            'Accept: application/json',
            'User-Agent: ' . $this->options->userAgent,
        ], $this->options->headers);

        $opts = [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HEADER         => false,
            CURLOPT_TIMEOUT        => $this->options->timeout,
            CURLOPT_CONNECTTIMEOUT => $this->options->connectTimeout,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_MAXREDIRS      => 5,
            CURLOPT_SSL_VERIFYPEER => true,
            CURLOPT_SSL_VERIFYHOST => 2,
        ];

        if ($this->options->apiKey !== null && $this->options->apiKey !== '') {
            $headers[] = 'Authorization: Bearer ' . $this->options->apiKey;
        }

        if ($method === 'POST' && $payload !== null) {
            $json = json_encode($payload, JSON_UNESCAPED_SLASHES);
            if ($json === false) {
                curl_close($ch);
                return (object) [
                    'data'       => [],
                    'error'      => 'Failed to encode JSON payload: ' . json_last_error_msg(),
                    'httpStatus' => null,
                ];
            }
            $opts[CURLOPT_POST]       = true;
            $opts[CURLOPT_POSTFIELDS] = $json;
            $headers[] = 'Content-Type: application/json';
            $headers[] = 'Content-Length: ' . strlen($json);
        }

        $opts[CURLOPT_HTTPHEADER] = $headers;
        curl_setopt_array($ch, $opts);

        // Retry loop with exponential backoff for 429 and 5xx responses.
        $attempt = 0;
        while (true) {
            $attempt++;
            $body     = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);

            // Network-level failure - not retryable (no HTTP status available).
            if ($body === false) {
                $error = curl_error($ch);
                $errno = curl_errno($ch);
                curl_close($ch);
                return (object) [
                    'data'       => [],
                    'error'      => 'cURL error (' . $errno . '): ' . $error,
                    'httpStatus' => null,
                ];
            }

            $decoded = json_decode((string) $body, true);

            // Check if the HTTP status is retryable (429 or 5xx).
            $isRetryable = $httpCode === 429 || $httpCode >= 500;

            if ($isRetryable && $attempt <= self::MAX_RETRIES) {
                $this->sleepBackoff($attempt);
                continue;
            }

            curl_close($ch);

            if (!is_array($decoded)) {
                return (object) [
                    'data'       => [],
                    'error'      => 'Invalid JSON response (HTTP ' . $httpCode . ')',
                    'httpStatus' => (int) $httpCode,
                ];
            }

            return (object) [
                'data'       => $decoded,
                'error'      => null,
                'httpStatus' => (int) $httpCode,
            ];
        }
    }

    /**
     * Sleep for an exponentially increasing duration before a retry attempt.
     *
     * Uses a base delay of BASE_BACKOFF_MS with exponential doubling:
     * attempt 1 → 200ms, attempt 2 → 400ms, attempt 3 → 800ms.
     *
     * @param int $attempt The current attempt number (1-based).
     */
    private function sleepBackoff(int $attempt): void
    {
        $delayMs = self::BASE_BACKOFF_MS * (2 ** ($attempt - 1));
        usleep($delayMs * 1000);
    }
}
